File: checkers.js
// checkers.js
// Версия на JavaScript с использованием классов, readline, цветной вывод, сохранение/загрузка (JSON)

const readline = require('readline');
const fs = require('fs');
const { randomInt } = require('crypto');

// Цвета ANSI
const colors = {
    reset: '\x1b[0m',
    black: '\x1b[30m',
    red: '\x1b[31m',
    green: '\x1b[32m',
    yellow: '\x1b[33m',
    blue: '\x1b[34m',
    magenta: '\x1b[35m',
    cyan: '\x1b[36m',
    white: '\x1b[37m',
    bg_black: '\x1b[40m',
    bg_white: '\x1b[47m',
    bg_red: '\x1b[41m',
    bg_green: '\x1b[42m',
    bg_yellow: '\x1b[43m',
    bg_blue: '\x1b[44m',
    bg_magenta: '\x1b[45m',
    bg_cyan: '\x1b[46m',
};

function colorize(text, fg = 'reset', bg = null) {
    const bgCode = bg ? colors[`bg_${bg}`] || '' : '';
    return `${colors[fg] || ''}${bgCode}${text}${colors.reset}`;
}

const ColorType = { WHITE: 1, BLACK: 2 };
const PieceType = { MAN: 1, KING: 2 };

class Piece {
    constructor(color, type) {
        this.color = color;
        this.type = type;
    }
}

class Board {
    static SIZE = 10;
    constructor() {
        this.grid = Array.from({ length: Board.SIZE }, () => Array(Board.SIZE).fill(null));
        this._initBoard();
    }

    _initBoard() {
        for (let r = 0; r < Board.SIZE; r++) {
            for (let c = 0; c < Board.SIZE; c++) {
                if ((r + c) % 2 === 1) {
                    if (r < 4) this.grid[r][c] = new Piece(ColorType.WHITE, PieceType.MAN);
                    else if (r > 5) this.grid[r][c] = new Piece(ColorType.BLACK, PieceType.MAN);
                }
            }
        }
    }

    getPiece(r, c) {
        if (this._inBounds(r, c)) return this.grid[r][c];
        return null;
    }

    _inBounds(r, c) {
        return r >= 0 && r < Board.SIZE && c >= 0 && c < Board.SIZE;
    }

    isOccupied(r, c) {
        return this.getPiece(r, c) !== null;
    }

    placePiece(r, c, piece) {
        this.grid[r][c] = piece;
    }

    removePiece(r, c) {
        this.grid[r][c] = null;
    }

    findPieces(color) {
        const result = [];
        for (let r = 0; r < Board.SIZE; r++) {
            for (let c = 0; c < Board.SIZE; c++) {
                const p = this.grid[r][c];
                if (p && p.color === color) result.push([r, c]);
            }
        }
        return result;
    }

    isKingRow(r, color) {
        return (color === ColorType.WHITE && r === Board.SIZE - 1) ||
               (color === ColorType.BLACK && r === 0);
    }

    promoteToKing(r, c) {
        const p = this.grid[r][c];
        if (p && p.type === PieceType.MAN) {
            this.grid[r][c] = new Piece(p.color, PieceType.KING);
        }
    }

    clone() {
        const b = new Board();
        for (let r = 0; r < Board.SIZE; r++) {
            for (let c = 0; c < Board.SIZE; c++) {
                const p = this.grid[r][c];
                b.grid[r][c] = p ? new Piece(p.color, p.type) : null;
            }
        }
        return b;
    }
}

class Move {
    constructor(fromR, fromC, toR, toC, captures) {
        this.fromRow = fromR;
        this.fromCol = fromC;
        this.toRow = toR;
        this.toCol = toC;
        this.captures = captures || []; // массив [r,c]
    }
}

class CheckersGame {
    constructor(twoPlayers = false) {
        this.board = new Board();
        this.currentTurn = ColorType.WHITE;
        this.twoPlayers = twoPlayers;
        this.moveHistory = [];
    }

    getOpponent(color) {
        return color === ColorType.WHITE ? ColorType.BLACK : ColorType.WHITE;
    }

    getAllValidMoves(color) {
        const pieces = this.board.findPieces(color);
        let moves = [];
        for (const [r, c] of pieces) {
            const piece = this.board.getPiece(r, c);
            if (piece.type === PieceType.MAN) {
                moves = moves.concat(this._getManMoves(r, c, color));
            } else {
                moves = moves.concat(this._getKingMoves(r, c, color));
            }
        }
        // Обязательное взятие
        const captures = moves.filter(m => m.captures.length > 0);
        if (captures.length > 0) return captures;
        return moves;
    }

    _getManMoves(row, col, color) {
        const moves = [];
        const dirs = color === ColorType.WHITE ? [[1,-1],[1,1]] : [[-1,-1],[-1,1]];
        for (const [dr, dc] of dirs) {
            const nr = row + dr, nc = col + dc;
            if (this.board._inBounds(nr, nc) && !this.board.isOccupied(nr, nc)) {
                moves.push(new Move(row, col, nr, nc, []));
            }
        }
        // Взятие (все направления)
        for (const dr of [-1,1]) {
            for (const dc of [-1,1]) {
                const nr = row + dr, nc = col + dc;
                if (this.board._inBounds(nr, nc)) {
                    const target = this.board.getPiece(nr, nc);
                    if (target && target.color !== color) {
                        const jr = nr + dr, jc = nc + dc;
                        if (this.board._inBounds(jr, jc) && !this.board.isOccupied(jr, jc)) {
                            moves.push(new Move(row, col, jr, jc, [[nr, nc]]));
                        }
                    }
                }
            }
        }
        return moves;
    }

    _getKingMoves(row, col, color) {
        const moves = [];
        for (const dr of [-1,1]) {
            for (const dc of [-1,1]) {
                // Обычные ходы
                let r = row + dr, c = col + dc;
                while (this.board._inBounds(r, c)) {
                    if (this.board.isOccupied(r, c)) break;
                    moves.push(new Move(row, col, r, c, []));
                    r += dr; c += dc;
                }
                // Взятия
                r = row + dr; c = col + dc;
                while (this.board._inBounds(r, c)) {
                    if (this.board.isOccupied(r, c)) {
                        const target = this.board.getPiece(r, c);
                        if (target.color !== color) {
                            let br = r + dr, bc = c + dc;
                            while (this.board._inBounds(br, bc)) {
                                if (this.board.isOccupied(br, bc)) break;
                                moves.push(new Move(row, col, br, bc, [[r, c]]));
                                break;
                            }
                        }
                        break;
                    }
                    r += dr; c += dc;
                }
            }
        }
        return moves;
    }

    isValidMove(move) {
        const all = this.getAllValidMoves(this.currentTurn);
        for (const m of all) {
            if (m.fromRow === move.fromRow && m.fromCol === move.fromCol &&
                m.toRow === move.toRow && m.toCol === move.toCol) {
                // Сравниваем captures (независимо от порядка)
                if (m.captures.length === move.captures.length) {
                    const set1 = new Set(m.captures.map(([r,c]) => `${r},${c}`));
                    const set2 = new Set(move.captures.map(([r,c]) => `${r},${c}`));
                    if (set1.size === set2.size && [...set1].every(k => set2.has(k))) {
                        return true;
                    }
                }
            }
        }
        return false;
    }

    makeMove(move) {
        if (!this.isValidMove(move)) return false;
        const piece = this.board.getPiece(move.fromRow, move.fromCol);
        if (!piece) return false;
        this.board.removePiece(move.fromRow, move.fromCol);
        this.board.placePiece(move.toRow, move.toCol, piece);
        for (const [r, c] of move.captures) {
            this.board.removePiece(r, c);
        }
        if (this.board.isKingRow(move.toRow, piece.color) && piece.type === PieceType.MAN) {
            this.board.promoteToKing(move.toRow, move.toCol);
        }
        this.moveHistory.push(move);
        this.currentTurn = this.getOpponent(this.currentTurn);
        return true;
    }

    isGameOver() {
        const white = this.board.findPieces(ColorType.WHITE);
        const black = this.board.findPieces(ColorType.BLACK);
        if (white.length === 0) return ColorType.BLACK;
        if (black.length === 0) return ColorType.WHITE;
        const moves = this.getAllValidMoves(this.currentTurn);
        if (moves.length === 0) return this.getOpponent(this.currentTurn);
        return null;
    }

    parseMove(input) {
        const parts = input.trim().split(/\s+/);
        if (parts.length < 2) return null;
        const coords = [];
        for (const p of parts) {
            const m = p.match(/^([A-Ja-j])([1-9]|10)$/);
            if (!m) return null;
            const col = m[1].toUpperCase().charCodeAt(0) - 'A'.charCodeAt(0);
            const row = parseInt(m[2]) - 1;
            coords.push([row, col]);
        }
        if (coords.length < 2) return null;
        const from = coords[0];
        const to = coords[coords.length - 1];
        const captures = [];
        for (let i = 0; i < coords.length - 1; i++) {
            const [r1, c1] = coords[i];
            const [r2, c2] = coords[i+1];
            if (Math.abs(r2-r1) !== Math.abs(c2-c1)) return null;
            const dr = r2 > r1 ? 1 : -1;
            const dc = c2 > c1 ? 1 : -1;
            let cr = r1 + dr, cc = c1 + dc;
            while (cr !== r2 || cc !== c2) {
                const piece = this.board.getPiece(cr, cc);
                if (!piece || piece.color === this.currentTurn) return null;
                captures.push([cr, cc]);
                cr += dr; cc += dc;
            }
        }
        return new Move(from[0], from[1], to[0], to[1], captures);
    }

    display(useColor = true) {
        const size = Board.SIZE;
        console.log('  ' + Array.from({length: size}, (_,i) => String.fromCharCode(65+i)).join(' '));
        for (let r = size - 1; r >= 0; r--) {
            const label = (r+1).toString().padStart(2, ' ');
            process.stdout.write(label + ' ');
            for (let c = 0; c < size; c++) {
                const piece = this.board.getPiece(r, c);
                const bg = (r + c) % 2 === 0 ? 'white' : 'black';
                if (!piece) {
                    if (useColor) {
                        process.stdout.write(colorize(' . ', 'reset', bg));
                    } else {
                        process.stdout.write(' . ');
                    }
                } else {
                    const symbol = piece.color === ColorType.WHITE ? 'W' : 'B';
                    const type = piece.type === PieceType.KING ? '*' : ' ';
                    const fg = bg === 'white' ? 'black' : 'white';
                    if (useColor) {
                        process.stdout.write(colorize(symbol + type, fg, bg));
                    } else {
                        process.stdout.write(symbol + type + ' ');
                    }
                }
            }
            console.log();
        }
        console.log(`Ход: ${this.currentTurn === ColorType.WHITE ? 'Белые' : 'Чёрные'}`);
    }

    getAIMove() {
        const moves = this.getAllValidMoves(this.currentTurn);
        if (moves.length === 0) return null;
        const captures = moves.filter(m => m.captures.length > 0);
        if (captures.length > 0) return captures[randomInt(captures.length)];
        return moves[randomInt(moves.length)];
    }

    saveState(filename) {
        const state = {
            board: this.board.grid.map(row => row.map(p => p ? {color: p.color, type: p.type} : null)),
            turn: this.currentTurn,
            history: this.moveHistory.map(m => ({fromRow: m.fromRow, fromCol: m.fromCol, toRow: m.toRow, toCol: m.toCol, captures: m.captures}))
        };
        fs.writeFileSync(filename, JSON.stringify(state, null, 2));
    }

    loadState(filename) {
        const data = JSON.parse(fs.readFileSync(filename, 'utf-8'));
        const board = new Board();
        for (let r = 0; r < Board.SIZE; r++) {
            for (let c = 0; c < Board.SIZE; c++) {
                const pData = data.board[r][c];
                board.grid[r][c] = pData ? new Piece(pData.color, pData.type) : null;
            }
        }
        this.board = board;
        this.currentTurn = data.turn;
        this.moveHistory = data.history.map(h => new Move(h.fromRow, h.fromCol, h.toRow, h.toCol, h.captures));
    }
}

function main() {
    const args = process.argv.slice(2);
    let useColor = true;
    let twoPlayers = false;
    let saveFile = null;
    let loadFile = null;
    for (let i = 0; i < args.length; i++) {
        if (args[i] === '--no-color') useColor = false;
        else if (args[i] === '--color') useColor = true;
        else if (args[i] === '--two-players') twoPlayers = true;
        else if (args[i] === '--save' && i+1 < args.length) saveFile = args[++i];
        else if (args[i] === '--load' && i+1 < args.length) loadFile = args[++i];
    }

    const game = new CheckersGame(twoPlayers);
    if (loadFile) game.loadState(loadFile);

    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    function promptMove() {
        game.display(useColor);
        const winner = game.isGameOver();
        if (winner) {
            console.log(`Победили ${winner === ColorType.WHITE ? 'Белые' : 'Чёрные'}!`);
            rl.close();
            return;
        }
        if (game.twoPlayers) {
            const name = game.currentTurn === ColorType.WHITE ? 'Белые' : 'Чёрные';
            rl.question(`Ход ${name}. Введите ход: `, handleInput);
        } else {
            if (game.currentTurn === ColorType.BLACK) {
                // Компьютер
                const move = game.getAIMove();
                if (move) {
                    game.makeMove(move);
                    console.log(`Компьютер сходил: ${String.fromCharCode(65+move.fromCol)}${move.fromRow+1} -> ${String.fromCharCode(65+move.toCol)}${move.toRow+1}`);
                    if (saveFile) game.saveState(saveFile);
                    promptMove();
                } else {
                    console.log('Компьютер не может ходить.');
                    rl.close();
                }
            } else {
                rl.question('Ваш ход (например, a1 b2): ', handleInput);
            }
        }
    }

    function handleInput(input) {
        if (input.toLowerCase() === 'quit' || input.toLowerCase() === 'exit') {
            rl.close();
            return;
        }
        const move = game.parseMove(input);
        if (!move) {
            console.log('Неверный формат хода.');
            promptMove();
            return;
        }
        if (game.makeMove(move)) {
            console.log('Ход принят.');
            if (saveFile) game.saveState(saveFile);
            promptMove();
        } else {
            console.log('Недопустимый ход.');
            promptMove();
        }
    }

    promptMove();
}

if (require.main === module) {
    main();
}

File: checkers.go
// checkers.go
// Версия на Go с использованием структур, методов, цветного вывода, сохранения/загрузки JSON

package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

// Цвета ANSI
const (
	reset  = "\033[0m"
	black  = "\033[30m"
	red    = "\033[31m"
	green  = "\033[32m"
	yellow = "\033[33m"
	blue   = "\033[34m"
	magenta = "\033[35m"
	cyan   = "\033[36m"
	white  = "\033[37m"
	bgBlack = "\033[40m"
	bgWhite = "\033[47m"
)

func colorize(text string, fg string, bg string) string {
	bgCode := ""
	if bg == "black" {
		bgCode = bgBlack
	} else if bg == "white" {
		bgCode = bgWhite
	}
	return fg + bgCode + text + reset
}

type ColorType int
const (
	WHITE ColorType = 1
	BLACK ColorType = 2
)

type PieceType int
const (
	MAN PieceType = 1
	KING PieceType = 2
)

type Piece struct {
	Color ColorType
	Type  PieceType
}

type Move struct {
	FromRow  int
	FromCol  int
	ToRow    int
	ToCol    int
	Captures [][2]int
}

type Board struct {
	grid [10][10]*Piece
}

func NewBoard() *Board {
	b := &Board{}
	for r := 0; r < 10; r++ {
		for c := 0; c < 10; c++ {
			if (r+c)%2 == 1 {
				if r < 4 {
					b.grid[r][c] = &Piece{WHITE, MAN}
				} else if r > 5 {
					b.grid[r][c] = &Piece{BLACK, MAN}
				}
			}
		}
	}
	return b
}

func (b *Board) inBounds(r, c int) bool {
	return r >= 0 && r < 10 && c >= 0 && c < 10
}

func (b *Board) GetPiece(r, c int) *Piece {
	if b.inBounds(r, c) {
		return b.grid[r][c]
	}
	return nil
}

func (b *Board) IsOccupied(r, c int) bool {
	return b.GetPiece(r, c) != nil
}

func (b *Board) PlacePiece(r, c int, p *Piece) {
	b.grid[r][c] = p
}

func (b *Board) RemovePiece(r, c int) {
	b.grid[r][c] = nil
}

func (b *Board) FindPieces(color ColorType) [][2]int {
	var res [][2]int
	for r := 0; r < 10; r++ {
		for c := 0; c < 10; c++ {
			if p := b.grid[r][c]; p != nil && p.Color == color {
				res = append(res, [2]int{r, c})
			}
		}
	}
	return res
}

func (b *Board) IsKingRow(r int, color ColorType) bool {
	return (color == WHITE && r == 9) || (color == BLACK && r == 0)
}

func (b *Board) PromoteToKing(r, c int) {
	if p := b.grid[r][c]; p != nil && p.Type == MAN {
		b.grid[r][c] = &Piece{p.Color, KING}
	}
}

func (b *Board) Clone() *Board {
	nb := NewBoard()
	for r := 0; r < 10; r++ {
		for c := 0; c < 10; c++ {
			if p := b.grid[r][c]; p != nil {
				nb.grid[r][c] = &Piece{p.Color, p.Type}
			} else {
				nb.grid[r][c] = nil
			}
		}
	}
	return nb
}

type Game struct {
	board       *Board
	currentTurn ColorType
	twoPlayers  bool
	moveHistory []Move
}

func NewGame(twoPlayers bool) *Game {
	return &Game{
		board:       NewBoard(),
		currentTurn: WHITE,
		twoPlayers:  twoPlayers,
		moveHistory: []Move{},
	}
}

func (g *Game) opponent(color ColorType) ColorType {
	if color == WHITE {
		return BLACK
	}
	return WHITE
}

func (g *Game) getManMoves(row, col int, color ColorType) []Move {
	var moves []Move
	dirs := [][2]int{}
	if color == WHITE {
		dirs = [][2]int{{1, -1}, {1, 1}}
	} else {
		dirs = [][2]int{{-1, -1}, {-1, 1}}
	}
	for _, d := range dirs {
		nr, nc := row+d[0], col+d[1]
		if g.board.inBounds(nr, nc) && !g.board.IsOccupied(nr, nc) {
			moves = append(moves, Move{row, col, nr, nc, [][2]int{}})
		}
	}
	for _, dr := range []int{-1, 1} {
		for _, dc := range []int{-1, 1} {
			nr, nc := row+dr, col+dc
			if g.board.inBounds(nr, nc) {
				target := g.board.GetPiece(nr, nc)
				if target != nil && target.Color != color {
					jr, jc := nr+dr, nc+dc
					if g.board.inBounds(jr, jc) && !g.board.IsOccupied(jr, jc) {
						moves = append(moves, Move{row, col, jr, jc, [][2]int{{nr, nc}}})
					}
				}
			}
		}
	}
	return moves
}

func (g *Game) getKingMoves(row, col int, color ColorType) []Move {
	var moves []Move
	for _, dr := range []int{-1, 1} {
		for _, dc := range []int{-1, 1} {
			// simple moves
			r, c := row+dr, col+dc
			for g.board.inBounds(r, c) {
				if g.board.IsOccupied(r, c) {
					break
				}
				moves = append(moves, Move{row, col, r, c, [][2]int{}})
				r += dr
				c += dc
			}
			// captures
			r, c = row+dr, col+dc
			for g.board.inBounds(r, c) {
				if g.board.IsOccupied(r, c) {
					target := g.board.GetPiece(r, c)
					if target != nil && target.Color != color {
						br, bc := r+dr, c+dc
						for g.board.inBounds(br, bc) {
							if g.board.IsOccupied(br, bc) {
								break
							}
							moves = append(moves, Move{row, col, br, bc, [][2]int{{r, c}}})
							break
						}
					}
					break
				}
				r += dr
				c += dc
			}
		}
	}
	return moves
}

func (g *Game) GetAllValidMoves(color ColorType) []Move {
	pieces := g.board.FindPieces(color)
	var moves []Move
	for _, p := range pieces {
		r, c := p[0], p[1]
		piece := g.board.GetPiece(r, c)
		if piece == nil {
			continue
		}
		if piece.Type == MAN {
			moves = append(moves, g.getManMoves(r, c, color)...)
		} else {
			moves = append(moves, g.getKingMoves(r, c, color)...)
		}
	}
	var captures []Move
	for _, m := range moves {
		if len(m.Captures) > 0 {
			captures = append(captures, m)
		}
	}
	if len(captures) > 0 {
		return captures
	}
	return moves
}

func (g *Game) IsValidMove(move Move) bool {
	all := g.GetAllValidMoves(g.currentTurn)
	for _, m := range all {
		if m.FromRow == move.FromRow && m.FromCol == move.FromCol &&
			m.ToRow == move.ToRow && m.ToCol == move.ToCol {
			if len(m.Captures) == len(move.Captures) {
				// compare sets
				set1 := make(map[[2]int]bool)
				for _, cap := range m.Captures {
					set1[cap] = true
				}
				set2 := make(map[[2]int]bool)
				for _, cap := range move.Captures {
					set2[cap] = true
				}
				ok := true
				for k := range set1 {
					if !set2[k] {
						ok = false
						break
					}
				}
				if ok {
					return true
				}
			}
		}
	}
	return false
}

func (g *Game) MakeMove(move Move) bool {
	if !g.IsValidMove(move) {
		return false
	}
	piece := g.board.GetPiece(move.FromRow, move.FromCol)
	if piece == nil {
		return false
	}
	g.board.RemovePiece(move.FromRow, move.FromCol)
	g.board.PlacePiece(move.ToRow, move.ToCol, piece)
	for _, cap := range move.Captures {
		g.board.RemovePiece(cap[0], cap[1])
	}
	if g.board.IsKingRow(move.ToRow, piece.Color) && piece.Type == MAN {
		g.board.PromoteToKing(move.ToRow, move.ToCol)
	}
	g.moveHistory = append(g.moveHistory, move)
	g.currentTurn = g.opponent(g.currentTurn)
	return true
}

func (g *Game) IsGameOver() *ColorType {
	white := g.board.FindPieces(WHITE)
	black := g.board.FindPieces(BLACK)
	if len(white) == 0 {
		res := BLACK
		return &res
	}
	if len(black) == 0 {
		res := WHITE
		return &res
	}
	moves := g.GetAllValidMoves(g.currentTurn)
	if len(moves) == 0 {
		res := g.opponent(g.currentTurn)
		return &res
	}
	return nil
}

func (g *Game) ParseMove(input string) *Move {
	parts := strings.Fields(input)
	if len(parts) < 2 {
		return nil
	}
	coords := [][2]int{}
	re := regexp.MustCompile(`^([A-Ja-j])([1-9]|10)$`)
	for _, p := range parts {
		m := re.FindStringSubmatch(p)
		if m == nil {
			return nil
		}
		col := int(m[1][0] - 'A')
		if m[1][0] >= 'a' {
			col = int(m[1][0] - 'a')
		}
		row, _ := strconv.Atoi(m[2])
		row--
		coords = append(coords, [2]int{row, col})
	}
	if len(coords) < 2 {
		return nil
	}
	from := coords[0]
	to := coords[len(coords)-1]
	var captures [][2]int
	for i := 0; i < len(coords)-1; i++ {
		r1, c1 := coords[i][0], coords[i][1]
		r2, c2 := coords[i+1][0], coords[i+1][1]
		if abs(r2-r1) != abs(c2-c1) {
			return nil
		}
		dr := 1
		if r2 < r1 {
			dr = -1
		}
		dc := 1
		if c2 < c1 {
			dc = -1
		}
		cr, cc := r1+dr, c1+dc
		for cr != r2 || cc != c2 {
			piece := g.board.GetPiece(cr, cc)
			if piece == nil || piece.Color == g.currentTurn {
				return nil
			}
			captures = append(captures, [2]int{cr, cc})
			cr += dr
			cc += dc
		}
	}
	return &Move{from[0], from[1], to[0], to[1], captures}
}

func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

func (g *Game) Display(useColor bool) {
	fmt.Print("  ")
	for c := 0; c < 10; c++ {
		fmt.Printf("%c ", 'A'+c)
	}
	fmt.Println()
	for r := 9; r >= 0; r-- {
		fmt.Printf("%2d ", r+1)
		for c := 0; c < 10; c++ {
			piece := g.board.GetPiece(r, c)
			bg := "black"
			if (r+c)%2 == 0 {
				bg = "white"
			}
			if piece == nil {
				if useColor {
					fmt.Print(colorize(" . ", reset, bg))
				} else {
					fmt.Print(" . ")
				}
			} else {
				sym := "W"
				if piece.Color == BLACK {
					sym = "B"
				}
				if piece.Type == KING {
					sym += "*"
				} else {
					sym += " "
				}
				fg := "black"
				if bg == "black" {
					fg = "white"
				}
				if useColor {
					fmt.Print(colorize(sym, fg, bg))
				} else {
					fmt.Print(sym + " ")
				}
			}
		}
		fmt.Println()
	}
	turn := "Белые"
	if g.currentTurn == BLACK {
		turn = "Чёрные"
	}
	fmt.Printf("Ход: %s\n", turn)
}

func (g *Game) GetAIMove() *Move {
	moves := g.GetAllValidMoves(g.currentTurn)
	if len(moves) == 0 {
		return nil
	}
	var captures []Move
	for _, m := range moves {
		if len(m.Captures) > 0 {
			captures = append(captures, m)
		}
	}
	if len(captures) > 0 {
		return &captures[rand.Intn(len(captures))]
	}
	return &moves[rand.Intn(len(moves))]
}

type serializedBoard [10][10]*PieceData
type PieceData struct {
	Color int `json:"color"`
	Type  int `json:"type"`
}
type serializedMove struct {
	FromRow  int        `json:"fromRow"`
	FromCol  int        `json:"fromCol"`
	ToRow    int        `json:"toRow"`
	ToCol    int        `json:"toCol"`
	Captures [][2]int   `json:"captures"`
}
type serializedGame struct {
	Board   [10][10]*PieceData `json:"board"`
	Turn    int                `json:"turn"`
	History []serializedMove   `json:"history"`
}

func (g *Game) SaveState(filename string) {
	sb := [10][10]*PieceData{}
	for r := 0; r < 10; r++ {
		for c := 0; c < 10; c++ {
			if p := g.board.grid[r][c]; p != nil {
				sb[r][c] = &PieceData{int(p.Color), int(p.Type)}
			} else {
				sb[r][c] = nil
			}
		}
	}
	hist := []serializedMove{}
	for _, m := range g.moveHistory {
		hist = append(hist, serializedMove{m.FromRow, m.FromCol, m.ToRow, m.ToCol, m.Captures})
	}
	sg := serializedGame{sb, int(g.currentTurn), hist}
	data, _ := json.MarshalIndent(sg, "", "  ")
	ioutil.WriteFile(filename, data, 0644)
}

func (g *Game) LoadState(filename string) {
	data, _ := ioutil.ReadFile(filename)
	var sg serializedGame
	json.Unmarshal(data, &sg)
	board := NewBoard()
	for r := 0; r < 10; r++ {
		for c := 0; c < 10; c++ {
			if sg.Board[r][c] != nil {
				board.grid[r][c] = &Piece{ColorType(sg.Board[r][c].Color), PieceType(sg.Board[r][c].Type)}
			} else {
				board.grid[r][c] = nil
			}
		}
	}
	g.board = board
	g.currentTurn = ColorType(sg.Turn)
	g.moveHistory = []Move{}
	for _, m := range sg.History {
		g.moveHistory = append(g.moveHistory, Move{m.FromRow, m.FromCol, m.ToRow, m.ToCol, m.Captures})
	}
}

func main() {
	rand.Seed(time.Now().UnixNano())
	args := os.Args[1:]
	useColor := true
	twoPlayers := false
	saveFile := ""
	loadFile := ""
	for i := 0; i < len(args); i++ {
		switch args[i] {
		case "--no-color":
			useColor = false
		case "--color":
			useColor = true
		case "--two-players":
			twoPlayers = true
		case "--save":
			if i+1 < len(args) {
				saveFile = args[i+1]
				i++
			}
		case "--load":
			if i+1 < len(args) {
				loadFile = args[i+1]
				i++
			}
		case "--help":
			fmt.Println("Использование: go run checkers.go [опции]")
			fmt.Println("  --color               Цветной вывод (по умолчанию)")
			fmt.Println("  --no-color            Отключить цвет")
			fmt.Println("  --two-players         Режим для двух игроков")
			fmt.Println("  --save <file>         Сохранять состояние")
			fmt.Println("  --load <file>         Загрузить состояние")
			return
		}
	}

	game := NewGame(twoPlayers)
	if loadFile != "" {
		game.LoadState(loadFile)
	}

	scanner := bufio.NewScanner(os.Stdin)

	for {
		game.Display(useColor)
		winner := game.IsGameOver()
		if winner != nil {
			if *winner == WHITE {
				fmt.Println("Победили Белые!")
			} else {
				fmt.Println("Победили Чёрные!")
			}
			return
		}
		if game.twoPlayers {
			name := "Белые"
			if game.currentTurn == BLACK {
				name = "Чёрные"
			}
			fmt.Printf("Ход %s. Введите ход: ", name)
			if !scanner.Scan() {
				return
			}
			input := scanner.Text()
			if input == "quit" || input == "exit" {
				return
			}
			move := game.ParseMove(input)
			if move == nil {
				fmt.Println("Неверный формат хода.")
				continue
			}
			if game.MakeMove(*move) {
				fmt.Println("Ход принят.")
				if saveFile != "" {
					game.SaveState(saveFile)
				}
			} else {
				fmt.Println("Недопустимый ход.")
			}
		} else {
			if game.currentTurn == BLACK {
				move := game.GetAIMove()
				if move == nil {
					fmt.Println("Компьютер не может ходить.")
					return
				}
				game.MakeMove(*move)
				fmt.Printf("Компьютер сходил: %c%d -> %c%d\n", 'A'+move.FromCol, move.FromRow+1, 'A'+move.ToCol, move.ToRow+1)
				if saveFile != "" {
					game.SaveState(saveFile)
				}
			} else {
				fmt.Print("Ваш ход (например, a1 b2): ")
				if !scanner.Scan() {
					return
				}
				input := scanner.Text()
				if input == "quit" || input == "exit" {
					return
				}
				move := game.ParseMove(input)
				if move == nil {
					fmt.Println("Неверный формат хода.")
					continue
				}
				if game.MakeMove(*move) {
					fmt.Println("Ход принят.")
					if saveFile != "" {
						game.SaveState(saveFile)
					}
				} else {
					fmt.Println("Недопустимый ход.")
				}
			}
		}
	}
}
